src/main/java/com/ldtteam/structurize/placement/handlers/placement/PlacementHandlers.java

```diff
@@ -63,6 +97,44 @@ private PlacementHandlers()
          */
     }
 
+    public static class FluidSubstitutionPlacementHandler implements IPlacementHandler
+    {
+        BlockState existing;
```

Comment:
I don't think this existing var here is a good idea at all.

Reply:
agree, never do that

Reply:
So I've been considering this throughout the day. I was aware it is an unorthadox solution, and originally I did it just so I could get testing. But I'm not really sure what the best alternative really is. I think this is a unique case where instead of removing something or ignoring it, the fluid block needs to take in the block that was already there so it can waterlog it.

I guess I could just inline all this into the main ``handleBlockPlacement``, but I was preferring to avoid putting exceptions in engine-like code (even though you already have some for the solid placeholder anyway).

Reply:
can't we query the existing block in the handle?

Reply:
the `canHandle` check should be just `instanceof BlockFluidSubstitution`

Reply:
oh I'm completely blind I did not see that ``worldPos`` was passed in to the handler. That could have saved me a bit of effort. I must've mixed it up for being a local position or something ... idk.

Reply:
=D

Reply:
OK I'm not as blind as I seem; I remember the reason I originally did it. The block is removed before the new one is handled, so querying by worldPos doesn't work because there's no longer a block there to waterlog. I'm thinking of just handling this as an exception inline, where I have access to the ``worldState`` variable.

Reply:
What we can do, might be a bit more effort, is also create in the handlers a removalHandler, and call that instead of the fixed onRemove in code, then all the existing handlers just use the default onRemoval in the interface, and the substitution blocks handle it themselves?

Reply:
That certainly does seem much more extensible. I like it. I'll have a go implementing that when my schedule frees up later this week.

Reply:
@ ShadowProtocol you should be able to adapt the calling of handler methods to include the old block, (try to use defaulting methods to make the diff smaller)